compiler/resolution.common.jvm/src/org/jetbrains/kotlin/resolve/jvm/KotlinJavaPsiFacade.java

```diff
@@ -353,24 +360,22 @@ public PsiPackage findPackage(@NotNull String qualifiedName, GlobalSearchScope s
                     }
                 }
 
-                cache.hasPackageInAllScopeCache.put(qualifiedName, found);
+                if (found || notFoundCacheType != NotFoundPackagesCachingStrategy.CacheType.NO_CACHING)
+                    cache.hasPackageInAllScopeCache.put(qualifiedName, found);
             }
         }
 
-        // qualifiedName could be like a proper package name, e.g `org.jetbrains.kotlin`
-        // but it could be as well part of typed text like `fooba`
-        //
-        // all those temporary names and those don't even look like a package name should be stored in a short term cache
-        // while names those are potentially proper package name could be stored for a long time
-        // (till PROJECT_ROOTS or specific VFS changes)
-        boolean packageLikeQName = qualifiedName.indexOf('.') > 0;
-
-        ConcurrentMap<Pair<String, GlobalSearchScope>, PsiPackage> notFoundPackageInScopeCache =
-                // store NULL_PACKAGE (attribute that package not found) in a long term cache if:
-                // - library search scope
-                // - qualifiedName looks like package (has `.` in its name)
-                isALibrarySearchScope && packageLikeQName ?
-                cache.packageInLibScopeCache : cache.packageInScopeCache;
+        ConcurrentMap<Pair<String, GlobalSearchScope>, PsiPackage> notFoundPackageInScopeCache;
+        switch (notFoundCacheType) {
+            case LIB_SCOPE:
+                notFoundPackageInScopeCache = cache.packageInLibScopeCache;
+                break;
+            case SCOPE:
+                notFoundPackageInScopeCache = cache.packageInScopeCache;
+                break;
+            default:
```

Comment:
I'd explicitly enumerated NO_CACHING here

Reply:
Handled this case